File: python/day_10/part_2.py
import re

from pulp import *


def solve_machine(buttons, joltages):
    n_buttons = len(buttons)
    n_counters = len(joltages)

    prob = LpProblem("MinimizeButtonPresses", LpMinimize)

    x = [LpVariable(f"button_{i}", lowBound=0, cat="Integer") for i in range(n_buttons)]

    prob += lpSum(x)

    for counter_idx in range(n_counters):
        counter_sum = lpSum(
            [
                x[btn_idx]
                for btn_idx, button in enumerate(buttons)
                if counter_idx in button
            ]
        )
        prob += counter_sum == joltages[counter_idx]

    prob.solve(PULP_CBC_CMD(msg=0))

    if prob.status == 1:  # Optimal solution found
        return int(sum([v.varValue for v in x]))

    return -1


def solve(input: str) -> int:
    lines = [line.strip() for line in input.strip().split("\n") if line.strip()]

    total_presses = 0

    for line in lines:
        buttons = []
        for match in re.finditer(r"\(([0-9,]+)\)", line):
            button = [int(x) for x in match.group(1).split(",")]
            buttons.append(button)

        joltage_match = re.search(r"\{([0-9,]+)\}", line)
        joltages = [int(x) for x in joltage_match.group(1).split(",")]

        min_presses = solve_machine(buttons, joltages)
        total_presses += min_presses

    return total_presses


if __name__ == "__main__":
    with open("inputs/day_10.txt", "r") as f:
        input_data = f.read()
        result = solve(input_data)
        print(result)

File: go/day_10/part_2.go
package main

import (
	"fmt"
	"math"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func solve_machine(buttons [][]int, joltages []int) int {
	n_buttons := len(buttons)
	n_counters := len(joltages)

	min_presses := math.MaxInt32

	max_tries := 1000000
	if n_buttons <= 10 {
		max_tries = 1 << (n_buttons * 4)
		if max_tries > 1000000 {
			max_tries = 1000000
		}
	}

	for attempt := 0; attempt < max_tries; attempt++ {
		x := make([]int, n_buttons)
		total := 0
		for i := range x {
			x[i] = attempt % 100
			attempt /= 100
			total += x[i]
		}

		if total >= min_presses {
			continue
		}

		valid := true
		for counter_idx := 0; counter_idx < n_counters; counter_idx++ {
			counter_sum := 0
			for btn_idx := 0; btn_idx < n_buttons; btn_idx++ {
				found := false
				for _, c := range buttons[btn_idx] {
					if c == counter_idx {
						found = true
						break
					}
				}
				if found {
					counter_sum += x[btn_idx]
				}
			}
			if counter_sum != joltages[counter_idx] {
				valid = false
				break
			}
		}

		if valid && total < min_presses {
			min_presses = total
		}
	}

	if min_presses == math.MaxInt32 {
		return -1
	}
	return min_presses
}

func solve(input string) int {
	lines := []string{}
	for _, line := range strings.Split(strings.TrimSpace(input), "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			lines = append(lines, line)
		}
	}

	total_presses := 0

	for _, line := range lines {
		buttons := [][]int{}
		button_re := regexp.MustCompile(`\(([0-9,]+)\)`)
		for _, match := range button_re.FindAllStringSubmatch(line, -1) {
			button := []int{}
			for _, x_str := range strings.Split(match[1], ",") {
				x, _ := strconv.Atoi(x_str)
				button = append(button, x)
			}
			buttons = append(buttons, button)
		}

		joltage_re := regexp.MustCompile(`\{([0-9,]+)\}`)
		joltage_match := joltage_re.FindStringSubmatch(line)
		joltages := []int{}
		if joltage_match != nil {
			for _, x_str := range strings.Split(joltage_match[1], ",") {
				x, _ := strconv.Atoi(x_str)
				joltages = append(joltages, x)
			}
		}

		min_presses := solve_machine(buttons, joltages)
		total_presses += min_presses
	}

	return total_presses
}

func main() {
	input, err := os.ReadFile("../inputs/day_10.txt")
	if err != nil {
		panic(err)
	}
	result := solve(string(input))
	fmt.Println(result)
}
